scripts/prepdocslib/textsplitter.py

```diff
@@ -16,16 +18,92 @@ def split_pages(self, pages: List[Page]) -> Generator[SplitPage, None, None]:
             yield  # pragma: no cover - this is necessary for mypy to type check
 
 
+STANDARD_LINE_BREAKS = [",", ";", ":", " ", "(", ")", "[", "]", "{", "}", "\t", "\n"]
+CJK_LINE_BREAKS = [
+    "。",
+    "、",
+    "，",
+    "；",
+    "：",
+    "（",
+    "）",
+    "【",
+    "】",
+    "「",
+    "」",
+    "『",
+    "』",
+    "〔",
+    "〕",
+    "〈",
+    "〉",
+    "《",
+    "》",
+    "〖",
+    "〗",
+    "〘",
+    "〙",
+    "〚",
+    "〛",
+    "〝",
+    "〞",
+    "〟",
+    "〰",
+    "〾",
+    "〿",
+    "–",
+    "—",
+    "‘",
+    "’",
+    "‚",
+    "‛",
+    "“",
+    "”",
+    "„",
+    "‟",
+    "‹",
+    "›",
+    "‽",
+    "⁇",
+    "⁈",
+    "⁉",
+    "⸮",
+    "ⸯ",
+    "。",
+    "々",
+    "〻",
+    "〼",
+]
+cl100k = tiktoken.get_encoding("cl100k_base")
+
+
+def split_page_by_max_tokens(page_num: int, text: str, max_tokens: int) -> Generator[SplitPage, None, None]:
+    """
+    Splits page by tokens
+    """
+    tokens = cl100k.encode(text)
+    if len(tokens) < max_tokens:
+        yield SplitPage(page_num=page_num, text=text)
+    else:
+        # Split page in half and call function again
+        # Overlap first and second halves by 5%
+        first_half = text[: int(len(text) // 2.05)]
+        second_half = text[int(len(text) // 1.95) :]
+        yield from split_page_by_max_tokens(page_num, first_half, max_tokens)
+        yield from split_page_by_max_tokens(page_num, second_half, max_tokens)
+
+
 class SentenceTextSplitter(TextSplitter):
     """
     Class that splits pages into smaller chunks. This is required because embedding models may not be able to analyze an entire page at once
     """
 
-    def __init__(self, has_image_embeddings: bool, verbose: bool = False):
-        self.sentence_endings = [".", "!", "?"]
-        self.word_breaks = [",", ";", ":", " ", "(", ")", "[", "]", "{", "}", "\t", "\n"]
+    def __init__(self, has_image_embeddings: bool, verbose: bool = False, max_tokens_per_section: int = 500):
+        self.sentence_endings = [".", "!", "?"] + ["。", "！", "？"]
+        self.word_breaks = STANDARD_LINE_BREAKS + CJK_LINE_BREAKS
```

Comment:
It confuses me that the constants are called LINE_BREAKS but the variable name is called word_breaks. Which is it? A comment may help as well.